File: hphp/hack/src/hhbc/generator.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

module T = Aast

let generator_visitor () =
  (* is_generator, is_pair_generator *)
  let state = ref (false, false) in
  object
    inherit [_] Aast.iter

    method state () = !state

    method! on_Yield () afield =
      match afield with
      | T.AFvalue _ ->
        let (_, is_pair_generator) = !state in
        state := (true, is_pair_generator)
      | T.AFkvalue (_, _) -> state := (true, true)

    method! on_Yield_break _ =
      let (_, is_pair_generator) = !state in
      state := (true, is_pair_generator)

    method! on_Yield_from _ _ =
      let (_, is_pair_generator) = !state in
      state := (true, is_pair_generator)

    method! on_class_ _ _ = ()

    method! on_fun_ _ _ = ()
  end

(* Returns a tuple of is_generator and is_pair_generator *)
let is_function_generator b =
  let visitor = generator_visitor () in
  let _ = visitor#on_program () b in
  visitor#state ()

File: hphp/hack/src/hhbc/generator.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use ast_body::AstBody;
use oxidized::{
    aast_visitor::{visit, AstParams, Node, Visitor},
    ast as tast,
};

pub fn is_function_generator(body: &AstBody) -> (bool, bool) {
    struct S((bool, bool));

    impl<'ast> Visitor<'ast> for S {
        type P = AstParams<(), ()>;

        fn object(&mut self) -> &mut dyn Visitor<'ast, P = Self::P> {
            self
        }

        fn visit_expr_(&mut self, c: &mut (), expr_: &tast::Expr_) -> Result<(), ()> {
            match expr_ {
                tast::Expr_::Yield(afield) => Ok(match afield.as_ref() {
                    tast::Afield::AFvalue(_) => (self.0).0 = true,
                    tast::Afield::AFkvalue(_, _) => self.0 = (true, true),
                }),
                tast::Expr_::YieldBreak => Ok((self.0).0 = true),
                _ => expr_.recurse(c, self),
            }
        }

        fn visit_class_(&mut self, _: &mut (), _: &tast::Class_) -> Result<(), ()> {
            Ok(())
        }
        fn visit_fun_(&mut self, _: &mut (), _: &tast::Fun_) -> Result<(), ()> {
            Ok(())
        }
    }
    let mut state = S((false, false));
    visit(&mut state, &mut (), body).unwrap();
    state.0
}
